### PR TITLE
Task/3731/cdist ssh key handler

### PR DESCRIPTION
**Pull request by pcoder**
This PR proposes cdist/celery based solution for managing user's public keys on hosts.
I left a few notes about this PR in redmine task 3731.

### CODE REVIEW

File: hosting/forms.py [hunk 1]
```diff
@@ -65,19 +71,47 @@ def clean_confirm_password(self):
...
+        openssh_pubkey = self.data.get('public_key')
+        data = None
+        try:
+            key_type, key_string, comment = openssh_pubkey.split()
+            data = base64.decodebytes(key_string.encode('utf-8'))
+        except Exception as e:
+            logger.error("Exception while decoding ssh key {}".format(e))
+            raise forms.ValidationError(KEY_ERROR_MESSAGE)
+        int_len = 4
+        str_len = struct.unpack('>I', data[:int_len])[0]
```
Comment (telmich) [inline hosting/forms.py]: Does this also accept elliptic curve based keys?

This is a valid public ssh key:

[20:33] manager:~% cat .ssh/id_ed25519.pub           
ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIE9P5aA5O4Vsgg5sB6Ojk+z1f5F97hvwRRd3gi+a+GGq nico-2016

File: opennebula_api/models.py [hunk 2]
```diff
@@ -208,6 +217,33 @@ def get_vm(self, vm_id):
...
+    def get_primary_ipv4(self, vm_id):
+        """
+        Returns the primary IPv4 of the given vm.
+        To be changed later.
+
+        :return: An IP address string, if it exists else returns None
+        """
+        all_ipv4s = self.get_vm_ipv4_addresses(vm_id)
+        if len(all_ipv4s) > 0:
+            return all_ipv4s[0]
```
Comment (telmich) [inline opennebula_api/models.py]: I suggest to even return the whole list as in " ".join(all_ipv4s)

File: hosting/forms.py (hunk 1 above)
Comment (pcoder) [reply hosting/forms.py]: That indeed does not work for ed25519 keys. I have updated the code to use sshpubkeys module that has a parse method to  parse a public key @telmich

File: opennebula_api/models.py (hunk 2 above)
Comment (pcoder) [reply opennebula_api/models.py]: We have `get_vm_ipv4_addresses` below for that purpose @telmich. I created the above method to identify which IP address to use to run the cdist config on.

**Comment by verysanghee** [comment]
This key exists already with the name "test" -> Der SSH-Key mit dem Name "test" existiert bereits.

**Pull request merged**